File: libs/chat/src/lib/a2ui/envelope-normalizer.ts
// libs/chat/src/lib/a2ui/envelope-normalizer.ts
// SPDX-License-Identifier: MIT

const ENVELOPE_KEYS = ['surfaceUpdate', 'beginRendering', 'dataModelUpdate', 'deleteSurface'] as const;

/**
 * The parent LLM may emit envelope-tool arguments in four shapes (observed in
 * the spike across gpt-5-mini and gpt-5): the canonical {envelopes: [...]},
 * a singular typo {envelope: [...]}, positional keys {0: env, 1: env, ...}
 * when the model treats the args as the array, or a flat single envelope.
 * This pure function maps all four into a canonical envelope list.
 *
 * Strict-mode tool binding (OpenAI) should eliminate the non-canonical
 * shapes in production, but the normalizer is the safety net.
 */
export function normalizeEnvelopeArgs(
  args: Record<string, unknown> | null | undefined,
): unknown[] | null {
  if (!args || typeof args !== 'object' || Array.isArray(args)) return null;

  // (a) canonical: { envelopes: [...] }
  if (Array.isArray((args as { envelopes?: unknown }).envelopes)) {
    return (args as { envelopes: unknown[] }).envelopes;
  }
  // (b) singular typo: { envelope: [...] }
  if (Array.isArray((args as { envelope?: unknown }).envelope)) {
    return (args as { envelope: unknown[] }).envelope;
  }
  const keys = Object.keys(args);
  if (keys.length === 0) return null;
  // (c) positional keys: { 0: env, 1: env, ... }
  if (keys.every((k) => /^\d+$/.test(k))) {
    return keys
      .map((k) => Number(k))
      .sort((a, b) => a - b)
      .map((k) => (args as Record<string, unknown>)[String(k)]);
  }
  // (d) flat single envelope: { surfaceUpdate: {...} } | { beginRendering: ... } | etc
  if (ENVELOPE_KEYS.some((k) => k in args)) {
    return [args];
  }
  return null;
}

File: examples/chat/python/src/streaming/envelope_normalizer.py
# SPDX-License-Identifier: MIT
"""Normalises the four envelope-args shapes the parent LLM may emit into
a canonical envelope list. Parity with libs/chat/src/lib/a2ui/envelope-normalizer.ts.

The spike (examples/chat/python/spike/parent_envelope_quality.py) observed
these shapes across gpt-5-mini and gpt-5; strict-mode tool binding should
eliminate the non-canonical ones in production, but this normalizer is
the safety net.
"""
from __future__ import annotations

from typing import Any

_ENVELOPE_KEYS = ("surfaceUpdate", "beginRendering", "dataModelUpdate", "deleteSurface")


def normalize_envelope_args(args: Any) -> list[dict] | None:
    """Return a canonical envelope list, or None if `args` is unrecognised."""
    if not isinstance(args, dict) or not args:
        return None
    # (a) canonical {envelopes: [...]}
    envelopes = args.get("envelopes")
    if isinstance(envelopes, list):
        return envelopes
    # (b) singular {envelope: [...]} typo
    envelope = args.get("envelope")
    if isinstance(envelope, list):
        return envelope
    keys = list(args.keys())
    # (c) positional keys {"0": env, "1": env, ...}
    if keys and all(isinstance(k, str) and k.isdigit() for k in keys):
        return [args[k] for k in sorted(keys, key=int)]
    # (d) flat single envelope
    if any(k in args for k in _ENVELOPE_KEYS):
        return [args]
    return None
